examples/platforms/simulation/radio.c

```diff
@@ -1321,3 +960,623 @@ otError otPlatRadioGetRegion(otInstance *aInstance, uint16_t *aRegionCode)
+    }
+
+    if (isTxDone)
+    {
+        platformTransmitReturn(aInstance, &sTransmitFrame, (isAck ? &sReceiveFrame : NULL), OT_ERROR_NONE);
+    }
+    else if (!isAck || sPromiscuous)
+    {
+        radioProcessFrame(aInstance);
+    }
+
+exit:
+    return;
+}
+
+void radioSendMessage(otInstance *aInstance)
+{
+#if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT && OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
+    if (sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeIeOffset != 0)
+    {
+        uint8_t *timeIe = sTransmitFrame.mPsdu + sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeIeOffset;
+        uint64_t time = (uint64_t)((int64_t)otPlatTimeGet() + sTransmitFrame.mInfo.mTxInfo.mIeInfo->mNetworkTimeOffset);
+
+        *timeIe = sTransmitFrame.mInfo.mTxInfo.mIeInfo->mTimeSyncSeq;
+
+        *(++timeIe) = (uint8_t)(time & 0xff);
+        for (uint8_t i = 1; i < sizeof(uint64_t); i++)
+        {
+            time        = time >> 8;
+            *(++timeIe) = (uint8_t)(time & 0xff);
+        }
+    }
+#endif // OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT && OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    if (sCslPeriod > 0 && !sTransmitFrame.mInfo.mTxInfo.mIsHeaderUpdated)
+    {
+        otMacFrameSetCslIe(&sTransmitFrame, (uint16_t)sCslPeriod, getCslPhase());
+    }
+#endif
+
+    sTransmitMessage.mChannel = sTransmitFrame.mChannel;
+
+    otEXPECT(radioProcessTransmitSecurity(&sTransmitFrame) == OT_ERROR_NONE);
+    otPlatRadioTxStarted(aInstance, &sTransmitFrame);
+    radioTransmit(&sTransmitMessage, &sTransmitFrame);
+
+    // Wait for simulator confirmation signal of radio transmission in virtual time mode.
+    sTxWait = true;
+exit:
+    return;
+}
+
+void setupTransmission(otInstance *aInstance)
+{
+    otError  error = OT_ERROR_NONE;
+    uint32_t delay;
+
+    otEXPECT_ACTION(!sRadioTransmitting, error = OT_ERROR_ALREADY);
+
+    radioSendMessage(aInstance);
+
+exit:
+    if (error == OT_ERROR_ALREADY)
+    {
+        // Radio is already transmiting, wait for it to end
+        if (sTransmittingUntil > otPlatTimeGet())
+        {
+            delay = (uint32_t)(sTransmittingUntil - otPlatTimeGet()) + 1;
+        }
+        else
+        {
+            delay = 10; // Arbitrary number to give simulater enough time to send confirmation signal;
+        }
+
+        otPlatAlarmMicroStartAt(aInstance, otPlatAlarmMicroGetNow(), delay);
+    }
+
+    return;
+}
+
+bool platformRadioTaskPending(void)
+{
+    return sAckTxDonePending;
+}
+
+void platformRadioTxDone(otInstance *aInstance, uint8_t pktSeq)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    otEXPECT(sLastReportedState == OT_RADIO_STATE_TRANSMIT);
+
+    sRadioTransmitting = false;
+    sTransmittingUntil = 0;
+
+    if (sAckTxDonePending)
+    {
+        sAckTxDonePending = false;
+        setRadioState(sState);
+    }
+    else if (otMacFrameGetSequence(&sTransmitFrame) == pktSeq)
+    {
+        // Radio sState keeps in OT_RADIO_STATE_RECEIVE even when sending an ack.
```

Comment:
Is it intended "sState keeps in OT_RADIO_STATE_TRANSMIT even when receiving an ack." ?
That seems to match better with the code below.

Reply:
Yes, it is intentional. The radio hardware is set to use Ack; therefore, there is Tx and Rx state involved that should be seen by the OT-NS. However, from the device's point of view, it is just transmitting. If you change this state to Rx, it will think that it can already transmit something else.

Reply:
Agree, that's why I made my comment: your // comment line seems to be saying it in reverse.

```suggestion
        // Radio sState keeps in OT_RADIO_STATE_TRANSMIT even when receiving an Ack.
```

So the radio of the node transmitting a data frame stays in OT_RADIO_STATE_TRANSMIT while waiting for the Ack, while the OTNS records the radio state of this node as "Receive" for the time that it is receiving the Ack.

Reply:
I see that you understood. For a minute I thought you were right, but this is rather confusing, indeed.

Observe line 1163: the last OTNS reported state to get in this function should be OT_RADIO_STATE_TRANSMIT:
```    otEXPECT(sLastReportedState == OT_RADIO_STATE_TRANSMIT);```

So, if a device is in OT_RADIO_STATE_RECEIVE and receives a packet, it might send an ACK. However, during this ACK transmission, it will not change its own state to "transmiting", instead It keeps on OT_RADIO_STATE_RECEIVE. So, for OTNS accuracy, we did already report a change anyway for this ACK at line 1221
```
// Radio sState keeps in OT_RADIO_STATE_RECEIVE even when sending an ack.
// For energy accuracy, we make a report without changing the sState.
reportRadioStatusToOtns(OT_RADIO_STATE_TRANSMIT);
```

So, now we just need to fix it back with line 1179
```
// Radio sState keeps in OT_RADIO_STATE_RECEIVE even when sending an ack.
// For energy accuracy, we make a report without changing the sState.
if (otMacFrameIsAckRequested(&sTransmitFrame))
{
    reportRadioStatusToOtns(OT_RADIO_STATE_RECEIVE);
}
```

Reply:
So just to understand the code, there are 2 situations in which platformRadioTxDone() can be called:

1. Rx node completed sending an Ack (sState = Rx , sLastReportedState = Tx)
2. Tx node completed sending a non-ack (data) frame  (sState = Tx , sLastReportedState = Tx)

In case 1 it will hit the top 'if' part of the code, since sAckTxDonePending was true. The transmitted Ack frame data is sitting in the sAckFrame buffer. The sTransmitFrame buffer holds unrelated data that shouldn't be touched.

```
if (sAckTxDonePending)
```

In case 2 it will hit the bottom  'else if' part of the code. In this case the data frame transmitted sits in the sTransmitFrame buffer so the code may refer to it:

```
else if (otMacFrameGetSequence(&sTransmitFrame) == pktSeq)
```

The comment is situated in the case 2 code block, where sState = Tx. I don't understand why the comment says "Radio sState keeps in OT_RADIO_STATE_RECEIVE" because that contradicts the assumption that sState = Tx at that point.

Maybe the comment could be moved up to the case 1 code block! Or maybe I miss some case here.

Reply:
> So just to understand the code, there are 2 situations in which platformRadioTxDone() can be called:
> 
> 1. Rx node completed sending an Ack (sState = Rx , sLastReportedState = Tx)
> 2. Tx node completed sending a non-ack (data) frame  (sState = Tx , sLastReportedState = Tx)

There are two cases, correctly identified here.

> In case 1 it will hit the top 'if' part of the code, since sAckTxDonePending was true. The transmitted Ack frame data is sitting in the sAckFrame buffer. The sTransmitFrame buffer holds unrelated data that shouldn't be touched.
> 
> ```
> if (sAckTxDonePending)
> ```

Correct. 
In this case:
sState => OT_RADIO_STATE_TRANSMIT
sLastReportedState => OT_RADIO_STATE_TRANSMIT

> In case 2 it will hit the bottom 'else if' part of the code. In this case the data frame transmitted sits in the sTransmitFrame buffer so the code may refer to it:
> 
> ```
> else if (otMacFrameGetSequence(&sTransmitFrame) == pktSeq)
> ```

Correct.
In this case:
sState => OT_RADIO_STATE_RECEIVE
sLastReportedState => OT_RADIO_STATE_TRANSMIT

> The comment is situated in the case 2 code block, where sState = Tx. I don't understand why the comment says "Radio sState keeps in OT_RADIO_STATE_RECEIVE" because that contradicts the assumption that sState = Tx at that point.
> 
> Maybe the comment could be moved up to the case 1 code block! Or maybe I miss some case here.

You got everything correct, but the comment is still correct too (maybe I could improve with more details too). It repeats what is written at line 1219.
It sits in the "else if" case because the ```sState``` is different from the ```sLastReportedState```, thus it needs to be fixed (since we correctly reported the radio status as Tx when the device sent an ACK.)

On the other hand, if the last ```sTransmitFrame``` packet does not require an ACK (else case), then ```sLastReportedState``` will be correctly changed right after this function, triggered by ```platformTransmitReturn(aInstance, &sTransmitFrame, NULL, OT_ERROR_NONE)```.